source/MaterialXGenOgsXml/GlslFragmentGenerator.cpp

```diff
@@ -0,0 +1,251 @@
+    emitScopeEnd(stage);
+
+    // Add function body
+    emitScopeBegin(stage);
+
+    if (graph.hasClassification(ShaderNode::Classification::CLOSURE))
+    {
+        // Handle the case where the graph is a direct closure.
+        // We don't support rendering closures without attaching 
+        // to a surface shader, so just output black.
+        emitLine("return vec3(0.0)", stage);
+    }
+    else
+    {
+        // Add all function calls
+        emitFunctionCalls(graph, context, stage);
+
+        // Emit final result
+        const ShaderOutput* outputConnection = outputSocket->getConnection();
+        if (outputConnection)
+        {
+            string finalOutput = outputConnection->getVariable();
+            const string& channels = outputSocket->getChannels();
+            if (!channels.empty())
+            {
+                finalOutput = _syntax->getSwizzledVariable(finalOutput, outputConnection->getType(), channels, outputSocket->getType());
+            }
+
+            if (graph.hasClassification(ShaderNode::Classification::SURFACE))
+            {
+                if (context.getOptions().hwTransparency)
+                {
+                    // emitLine("float outAlpha = clamp(1.0 - dot(" + finalOutput + ".transparency, vec3(0.3333)), 0.0, 1.0)", stage);
+                    emitLine("return " + finalOutput + ".color", stage);
+                }
+                else
+                {
+                    emitLine("return " + finalOutput + ".color", stage);
+                }
+            }
+            else
+            {
+                if (!outputSocket->getType()->isFloat3())
+                {
+                    toVec3(outputSocket->getType(), finalOutput);
+                }
+                emitLine("return " + finalOutput, stage);
+            }
+        }
+        else
+        {
+            string outputValue = outputSocket->getValue() ? _syntax->getValue(outputSocket->getType(), *outputSocket->getValue()) : _syntax->getDefaultValue(outputSocket->getType());
+            if (!outputSocket->getType()->isFloat3())
+            {
+                string finalOutput = outputSocket->getVariable() + "_tmp";
+                emitLine(_syntax->getTypeName(outputSocket->getType()) + " " + finalOutput + " = " + outputValue, stage);
+                toVec3(outputSocket->getType(), finalOutput);
+                emitLine("return " + finalOutput, stage);
+            }
+            else
+            {
+                emitLine("return " + outputValue, stage);
+            }
+        }
+    }
+
+    // End function
+    emitScopeEnd(stage);
+
+    // Replace all token identifiers with real names
+    replaceIdentifiers(_identifiers, stage);
+
+    return shader;
+}
+
+void GlslFragmentGenerator::toVec3(const TypeDesc* type, string& variable)
+{
+    if (type->isFloat2())
+    {
+        variable = "vec3(" + variable + ", 0.0)";
+    }
+    else if (type->isFloat4())
+    {
+        variable = variable + ".xyz";
```

Comment:
Should we use premultiplied color here if it's a color4 vs truncating?

Reply:
Good point, I can fix that.